scripts/src/errors.ts

```diff
@@ -0,0 +1,129 @@
+export type ApiError = {
+	code: number;
+	error: string;
+	message: string;
+};
+
+export const CODES = {
+	Unauthorized: {
+		MissingToken: 1,
+		InvalidToken: 2
+	},
+	NotFound: {
+		App: 1,
+		Offer: 2,
+		Order: 3,
+		PublicKey: 4
+	},
+	Conflict: {
+		ExternalOrderExhausted: 1
+	},
+	Locked: {
+		OfferCapReached: 1
+	},
+	UnprocessableEntity: {
+		OpenedOrdersOnly: 1,
+		OpenedOrdersUnreturnable: 2
+	},
+	BadRequest: {
+		UnknownSignInType: 1,
+		WrongJWTAlgorithm: 2
+	}
+};
+
+export class MarketplaceError extends Error {
+	public readonly title: string;
+	public readonly status: number; // http status code
+	public readonly code: number; // our own internal codes
+
+	constructor(status: number, index: number, title: string, message: string) {
+		super(message);
+		this.code = Number(status + "" + index);
+		this.title = title;
+		this.status = status;
+	}
+
+	public toJson(): ApiError {
+		return {
+			code: this.code,
+			error: this.title,
+			message: this.message
+		};
+	}
+
+	public toString(): string {
+		return JSON.stringify(this.toJson());
+	}
+}
+
+function UnauthorizedError(index: number, message: string) {
+	return new MarketplaceError(401, index, "Unauthorized Request", message);
+}
+
+export function MissingToken() {
+	return UnauthorizedError(CODES.Unauthorized.MissingToken, "Request missing token");
+}
+
+export function InvalidToken(token: string) {
+	return UnauthorizedError(CODES.Unauthorized.InvalidToken, `Invalid token: ${ token }`);
+}
+
+function NotFoundError(index: number, message: string) {
+	return new MarketplaceError(404, index, "Not Found", message);
+}
+
+export function NoSuchApp(id: string) {
+	return NotFoundError(CODES.NotFound.App, `No such app: ${ id }`);
+}
+
+export function NoSuchOffer(id: string) {
+	return NotFoundError(CODES.NotFound.Offer, `No such offer: ${ id }`);
+}
+
+export function NoSuchOrder(id: string) {
+	return NotFoundError(CODES.NotFound.Order, `No such order: ${ id }`);
+}
+
+export function NoSuchPublicKey(appId: string, keyid: string) {
+	return NotFoundError(CODES.NotFound.App, `Key "${ keyid }" not found for iss "${ appId }"`);
+}
+
+function ConflictError(index: number, message: string) {
+	return new MarketplaceError(409, index, "Conflict", message);
+}
+
+export function ExternalOrderExhausted() {
+	return ConflictError(CODES.Conflict.ExternalOrderExhausted, "User already completed offer, or has a pending order");
+}
+
+function LockedError(index: number, message: string) {
+	return new MarketplaceError(423, index, "Resource Locked", message);
+}
+
+export function OfferCapReached(id: string) {
+	return LockedError(CODES.Locked.OfferCapReached, `Cap reached for offer: ${ id }`);
```

Comment:
423 means the resource (File/Document/Object) is locked, there's no locking here but an exhausted resource this can be one of these:
1. 410 error because the offer is gone
2. a 403 since it's a limit on the user. 
3. a lot of services try to keep it simple and return a 404 (offer not found) but I don't think it's descriptive enough.

Reply:
There could be a resource locked error when locking is implemented. I agree that here it should be something else, I tend to 410

Reply:
i chose `locked` because the resource might be available later on, i.e.: someone opened an order but canceled it before finishing.
all of the other ideas you gave don't convey this.

Reply:
If caps are increased? or if the cap is hourly based?
Can 410 change back to 200? i.e. gone and came back?

Reply:
404